### ISSUE TITLE
feature request: Adding `LDAP_BIND_PW_FILE` to support docker secrets

### ISSUE THREAD

**Issue by paulheg** [docker-mailserver/docker-mailserver#3457]
### Context

I want to connect LDAP but want to use docker secrets.

### Description

As the title states I want to use *docker secrets* to inject passwords.
I have looked trough the supported environment variables but couldn't find anything.
Since my credentials are already stored inside a file for docker secrets I would like to have one source of truth, besides the other security benefits.

Possbile candidates would be:

- `LDAP_BIND_PW_FILE`
- `SASLAUTHD_LDAP_PASSWORD_FILE`

### Alternatives

I can write the passwords into a `.env` file :-1: 

### Applicable Users

For anyone using **openldap** and docker compose.

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute?

I would love to contribute and implement it my self, at least I can try.

**Comment by polarathene** [comment]
Just a heads-up, Docker Secrets is not properly implemented unless you're using Docker Swarm. You lose many of the benefits of the feature outside of Swarm, it's effectively just a volume bind mount on the file and doesn't adjust permissions or ownership unless you've done so on the host yourself, no encryption or access to the vault only files.

---

I don't think we support Docker Swarm officially, and there is secrets management with other container runtimes like k8s.

I am not against implementing the support of reading a secret from a file to get a secret from. Just keep in mind that if an attacker has compromised a container, they may still have a means to access the secret. Either by manipulating the shell scripts to hijack and acquire the value at runtime (or just reading the file directly), accessing the memory (tricky), or other means 😅 

---

ENV isn't ideal for secrets, I agree. It can leak in a variety of ways unintentionally, especially some software that logs ENV or reports it via debug info (_I think some web frameworks can do this initially until further configured_), I recall either PHP or Python (Django) would output such when encountering a failure.

If you just want to inject your ENV from a file source, you can do that IIRC. There are some software like [Teller](https://github.com/tellerops/teller#whale-easy-docker-environment) that support encrypting your secrets in a file (or other sources) that aren't specific to just Docker, and allows decrypting that for ENV injection. You may not be able to do that with `docker` or `docker compose` commands in a way that avoids leaking ENV, and instead need to use it in the `CMD` or `ENTRYPOINT` or when running a specific command in shell script within the container.

**Comment by paulheg** [comment]
Hi thanks for your input.

I have done some small changes to the `variables-stack.sh` https://github.com/paulheg/docker-mailserver/commit/a050ff19cf8450ae09894144de1edfa7dfbd45a7

I also looked into the test suite, I have to say I'm a bit overwhelmed since I don't have that much experience with shell scripting. My first guess is to add to https://github.com/docker-mailserver/docker-mailserver/blob/master/test/tests/serial/mail_with_ldap.bats.
On your contributing page it states you are in the progress of migrating all tests to `/paralell`.
I would love some guidance, as to how to properly test this.

BR
Paul

**Comment by polarathene** [comment]
> On your contributing page it states you are in the progress of migrating all tests to `/parallel`.
> I would love some guidance, as to how to properly test this.

Yes, and the LDAP test has not been converted yet. There is an existing LDAP PR open with a similar problem, so I'll try find some time to get this test converted so that you can both have an easier time.

Testing wise, you would include a file that could be used to test the new ENV path with, although I will point out that you have:

```bash
VARS[LDAP_BIND_PW]=$(<"${LDAP_BIND_PW_FILE}")
```

We export these ENV, so while they're not managed by Docker (_and thus I don't think Docker inspect will leak that ENV_), this does make visible elsewhere. For example anything that uses a bash shell (as root) will have this as part of their ENV, and the ENV will be present to various commands we run within our scripts:

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/scripts/startup/variables-stack.sh#L205-L222

For LDAP ENV, we have a script function we call that will look at all the ENV with the `LDAP_` prefix, and update a config file:

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/scripts/startup/setup.d/ldap.sh#L31

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/scripts/startup/setup.d/ldap.sh#L16-L23

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/scripts/helpers/utils.sh#L88

Each of those files listed above have a `bind_pw` key that will get the `LDAP_BIND_PW` ENV value. I haven't checked but permissions of those files or the two locations we put ENV files into may be readable by other users than `root`, which may make this entire endeavour provide minimal security benefit to you? (_or require further investigation_)

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/postfix/ldap-users.cf#L3

Additionally, these config files for LDAP are supported as user supplied, where we make a copy from the config volume to an internal location:

https://github.com/docker-mailserver/docker-mailserver/blob/b001f5a14056fa4df5bf50466247f3f4e4d8000e/target/scripts/startup/setup.d/ldap.sh#L7-L12

---

It's not great and no active maintainer is that familiar with LDAP or our integration beyond what I've referred you to above.

As for an actual test case, you could check that `LDAP_BIND_PW` becomes the value you expect from the file that should populate it, or check a LDAP `.cf` file from the list shown above. That's all the feature is doing right? We have quite a few tests that do exactly that for ENV config tests.

---

Is this still a feature you want to go ahead with, considering the security benefit may not be to the extent you originally wanted? (_or effort to accomplish that much higher_)

**Comment by polarathene** [comment]
@paulheg are you still wanting this feature btw?

I've updated our tests, and presently doing a refactor on LDAP support.

I could look into supporting this, but the password will ultimately end up in the `/etc/postfix/ldap/*.cf` files (_and the related Dovecot + saslauthd configs_).

The refactor allows you to provide your own files with just this setting, but since it'd be common among them all that'd be annoying and better served with ENV, so your request is appropriate.

I did look into `teller` as well, but it didn't seem a good fit for DMS.

---

I think instead of a separate ENV, we could probably check if it's a file (_using a valid filepath as an actual password would seem a bit silly_), and if so we'd just read that in.

**Comment by paulheg** [comment]
Hi @polarathene, 
thank you for your work :)

At the moment im sadly occupied but I still would like to see this feature.

I think it still would be a better idea to have a separate env variable, because a lot of strings are valid paths. I think even a random word would be a valid relative path?

Beat regards
Paul

**Comment by polarathene** [comment]
> I think even a random word would be a valid relative path?

Only if a file exists.

It's probably not great security in the first place, if your password would match a common file name or path inside the container 😅 

In that case, you'd find there is a problem with the LDAP auth, and if you look at the config generated in `/etc/postfix/ldap/*` the bind password would not be your unintentional relative filename but the contents of that file. Easy fix? 🤷‍♂️ 

Anyone with the ability to read the LDAP generated config, or the from the filepath assigned to the ENV would also still have access to that secret. It'll just be kept out of the ENV.

**Comment by aartoni** [comment]
Hello! I'd like to contribute this feature by implementing something similar to [what Grafana does in its entrypoint](https://github.com/grafana/grafana/blob/be426858529d9742808cc8ad4ed226379fb770da/packaging/docker/run.sh#L49-L61), that is, reading every env var ending in `__FILE`, then saving the content of the pointed files to the corresponding env var. Let me know if you like the idea!

**Comment by polarathene** [comment]
> then saving the content of the pointed files to the corresponding env var.
> Let me know if you like the idea!

Ok, I think I get the intent. These wouldn't appear in any ENV checks, only through the lifetime of the script as variables. They would likely get persisted into a file we write with our finalized ENV ([`/root/.bashrc` + `/etc/dms-settings`](https://github.com/docker-mailserver/docker-mailserver/blob/85793988d6d81ec21e19ccab96070fcd1f3d6208/target/scripts/startup/variables-stack.sh#L190-L207) which some of our utility scripts import, or any root bash shell would then have, in addition to config files that the secret is written to.

I think the ideal approach is usually to read the secret from the file when the secret is actually needed as input (like login credentials), but in the case of DMS which is integration glue the software AFAIK doesn't support this for the LDAP password, so what you propose is probably fine.

I don't have too much time to think about the topic, the other maintainers might raise a concern with it once you open a PR, but I think they'll be fine with it and only care about the script logic itself.

---

You may want to wait on https://github.com/docker-mailserver/docker-mailserver/pull/4323 as this seems like a method that'd belong to `variables-stack.sh`.

Otherwise feel free to raise a PR with the contribution and rebase afterwards if you're comfortable with that.
